### DISCUSSION TITLE
Unable to run orchard cms in docker with authentication

### DISCUSSION THREAD

**Discussion by jotacv** [OrchardCMS/OrchardCore#15900] [Q&A]
I'm having trouble dockerizing the app with the blank site scheme.

this is the log for a normal succesful run:

    jjcv@ltjjcv MINGW64 ~/WORKSPACES/netcore/NIHON/out
    $ dotnet nihon_core.dll
    Hosting environment: Production
    Content root path: C:\Users\JJCV\WORKSPACES\netcore\NIHON\out
    Now listening on: http://localhost:5000
    Now listening on: https://localhost:5001
    Application started. Press Ctrl+C to shut down.
    info: OrchardCore.Environment.Shell.ShellHost[0]
          Start creation of shells
    info: OrchardCore.Environment.Shell.Builders.ShellContextFactory[0]
          Creating shell context for tenant 'Default'
    info: OrchardCore.Environment.Shell.ShellHost[0]
          Done pre-creating and registering shells
    info: OrchardCore.Modules.ModularTenantRouterMiddleware[0]
          Begin Routing Request
    info: OrchardCore.Modules.ModularTenantRouterMiddleware[0]
          Begin Routing Request
    info: OrchardCore.DisplayManagement.Descriptors.DefaultShapeTableManager[0]
          Start building shape table
    info: OrchardCore.DisplayManagement.Descriptors.DefaultShapeTableManager[0]
          Start discovering shapes
    info: OrchardCore.DisplayManagement.Descriptors.DefaultShapeTableManager[0]
          Start discovering candidate views filenames
    info: OrchardCore.DisplayManagement.Descriptors.DefaultShapeTableManager[0]
          Done discovering candidate views filenames
    [... the same log entry like 50 times...]
    info: OrchardCore.DisplayManagement.Descriptors.DefaultShapeTableManager[0]
          Done discovering shapes
    info: OrchardCore.DisplayManagement.Descriptors.DefaultShapeTableManager[0]
          Done building shape table

The cms runs fine, and upon startup users are redirected to the Login page where they can choose to authenticate using their credentials or using Azure AD.

However, when running the same inside the container:

    warn: Microsoft.AspNetCore.DataProtection.Repositories.FileSystemXmlRepository[60]
          Storing keys in a directory '/root/.aspnet/DataProtection-Keys' that may not be persisted outside of the container. Protected data will be unavailable when container is destroyed.
    info: Microsoft.AspNetCore.DataProtection.KeyManagement.XmlKeyManager[0]
          User profile is available. Using '/root/.aspnet/DataProtection-Keys' as key repository; keys will not be encrypted at rest.
    Hosting environment: Development
    Content root path: /
    Now listening on: http://[::]:5000
    Application started. Press Ctrl+C to shut down.
    info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
          Request starting HTTP/1.1 GET http://localhost:5000/
    info: OrchardCore.Environment.Shell.ShellHost[0]
          Start creation of shells
    info: OrchardCore.Environment.Shell.ShellHost[0]
          Done pre-creating and registering shells
    info: OrchardCore.Modules.ModularTenantRouterMiddleware[0]
          Begin Routing Request
    info: Microsoft.AspNetCore.Routing.EndpointMiddleware[0]
          Executing endpoint 'OrchardCore.Setup.Controllers.SetupController.Index (OrchardCore.Setup)'
    warn: Microsoft.AspNetCore.DataProtection.Repositories.FileSystemXmlRepository[60]
          Storing keys in a directory '/App_Data/Sites/Default/DataProtection-Keys' that may not be persisted outside of the container. Protected data will be unavailable when container is destroyed.
    info: Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker[3]
          Route matched with {area = "OrchardCore.Setup", action = "Index", controller = "Setup"}. Executing controller action with signature System.Threading.Tasks.Task`1[Microsoft.AspNetCore.Mvc.ActionResult] Index(System.String) on controller OrchardCore.Setup.Controllers.SetupController (OrchardCore.Setup).
    info: Microsoft.AspNetCore.Authorization.DefaultAuthorizationService[2]
          Authorization failed.
    info: Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker[3]
          Authorization failed for the request at filter 'Microsoft.AspNetCore.Mvc.Authorization.AuthorizeFilter'.
    info: Microsoft.AspNetCore.Mvc.ChallengeResult[1]
          Executing ChallengeResult with authentication schemes ().
    info: Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker[2]
          Executed action OrchardCore.Setup.Controllers.SetupController.Index (OrchardCore.Setup) in 14.7009ms
    info: Microsoft.AspNetCore.Routing.EndpointMiddleware[1]
          Executed endpoint 'OrchardCore.Setup.Controllers.SetupController.Index (OrchardCore.Setup)'
    fail: Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware[1]
          An unhandled exception has occurred while executing the request.
    System.InvalidOperationException: No authenticationScheme was specified, and there was no DefaultChallengeScheme found. The default schemes can be set using either AddAuthentication(string defaultScheme) or AddAuthentication(Action<AuthenticationOptions    configureOptions).
       at Microsoft.AspNetCore.Authentication.AuthenticationService.ChallengeAsync(HttpContext context, String scheme, AuthenticationProperties properties)
       at Microsoft.AspNetCore.Mvc.ChallengeResult.ExecuteResultAsync(ActionContext context)
       at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAlwaysRunResultFilters>g__Awaited|26_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
       at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|19_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
       at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
       at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
       at Microsoft.AspNetCore.Localization.RequestLocalizationMiddleware.Invoke(HttpContext context)
       at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
       at OrchardCore.Modules.ModularTenantRouterMiddleware.Invoke(HttpContext httpContext) in C:\projects\orchardcore\src\OrchardCore\OrchardCore\Modules\ModularTenantRouterMiddleware.cs:line 82
       at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
       at OrchardCore.Environment.Shell.Scope.ShellScope.UsingAsync(Func`2 execute) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.Abstractions\Shell\Scope\ShellScope.cs:line 101
       at OrchardCore.Modules.ModularTenantContainerMiddleware.Invoke(HttpContext httpContext) in C:\projects\orchardcore\src\OrchardCore\OrchardCore\Modules\ModularTenantContainerMiddleware.cs:line 59
       at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
    info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
          Request finished in 1537.625ms 500 text/html; charset=utf-8
    info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
          Request starting HTTP/1.1 GET http://localhost:5000/
    info: OrchardCore.Modules.ModularTenantRouterMiddleware[0]
          Begin Routing Request
    info: Microsoft.AspNetCore.Routing.EndpointMiddleware[0]
          Executing endpoint 'OrchardCore.Setup.Controllers.SetupController.Index (OrchardCore.Setup)'
    info: Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker[3]
          Route matched with {area = "OrchardCore.Setup", action = "Index", controller = "Setup"}. Executing controller action with signature System.Threading.Tasks.Task`1[Microsoft.AspNetCore.Mvc.ActionResult] Index(System.String) on controller OrchardCore.Setup.Controllers.SetupController (OrchardCore.Setup).
    info: Microsoft.AspNetCore.Authorization.DefaultAuthorizationService[2]
          Authorization failed.
    info: Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker[3]
          Authorization failed for the request at filter 'Microsoft.AspNetCore.Mvc.Authorization.AuthorizeFilter'.
    info: Microsoft.AspNetCore.Mvc.ChallengeResult[1]
          Executing ChallengeResult with authentication schemes ().
    info: Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker[2]
          Executed action OrchardCore.Setup.Controllers.SetupController.Index (OrchardCore.Setup) in 1.9184ms
    info: Microsoft.AspNetCore.Routing.EndpointMiddleware[1]
          Executed endpoint 'OrchardCore.Setup.Controllers.SetupController.Index (OrchardCore.Setup)'
    fail: Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware[1]
          An unhandled exception has occurred while executing the request.
    System.InvalidOperationException: No authenticationScheme was specified, and there was no DefaultChallengeScheme found. The default schemes can be set using either AddAuthentication(string defaultScheme) or AddAuthentication(Action<AuthenticationOptions    configureOptions).
       at Microsoft.AspNetCore.Authentication.AuthenticationService.ChallengeAsync(HttpContext context, String scheme, AuthenticationProperties properties)
       at Microsoft.AspNetCore.Mvc.ChallengeResult.ExecuteResultAsync(ActionContext context)
       at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAlwaysRunResultFilters>g__Awaited|26_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
       at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|19_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
       at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
       at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
       at Microsoft.AspNetCore.Localization.RequestLocalizationMiddleware.Invoke(HttpContext context)
       at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
       at OrchardCore.Modules.ModularTenantRouterMiddleware.Invoke(HttpContext httpContext) in C:\projects\orchardcore\src\OrchardCore\OrchardCore\Modules\ModularTenantRouterMiddleware.cs:line 82
       at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
       at OrchardCore.Environment.Shell.Scope.ShellScope.UsingAsync(Func`2 execute) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.Abstractions\Shell\Scope\ShellScope.cs:line 101
       at OrchardCore.Modules.ModularTenantContainerMiddleware.Invoke(HttpContext httpContext) in C:\projects\orchardcore\src\OrchardCore\OrchardCore\Modules\ModularTenantContainerMiddleware.cs:line 59
       at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
    info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
          Request finished in 23.8062ms 500 text/html; charset=utf-8
    info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
          Request starting HTTP/1.1 GET http://localhost:5000/favicon.ico
    info: OrchardCore.Modules.ModularTenantRouterMiddleware[0]
          Begin Routing Request
    info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
          Request finished in 8.877600000000001ms 404

I have also changed the Startup.cs to require authenticated users on startup:

        public Startup(ILogger<Startup    logger, IConfiguration configuration)
        {
            Configuration = configuration;
            _logger = logger;
        }

        // This method gets called by the runtime. Use this method to add services to the container.
        // For more information on how to configure your application, visit https://go.microsoft.com/fwlink/?LinkID=398940
        public void ConfigureServices(IServiceCollection services)
        {
            services
                .AddOrchardCms()
                .AddOrchardCore();

            services.Configure<CookiePolicyOptions>(options =>
            {
                options.CheckConsentNeeded = context =    false;
                options.MinimumSameSitePolicy = SameSiteMode.None;
            });

            services.Configure<MvcOptions>(options =>
            {
                options.Filters.Add(new AuthorizeFilter(new AuthorizationPolicyBuilder()
                        .RequireAuthenticatedUser()
                        .Build()));
            });

            services.AddSession(opt =>
            {
                opt.Cookie.IsEssential = true;
            });

            services.AddScoped<IUserService, UserService>();
            services.AddScoped<IAuthenticationProvider, MsGraphAuthenticationProvider>();

        }

        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }
            else
            {
                app.UseExceptionHandler("/Home/Error");
                app.UseHsts();
            }

            app.UseStaticFiles();
            app.UseOrchardCore(builder =>
            {
                builder.UseCookiePolicy();
                builder.UseAuthentication();
            });
        }

At first i thought that orchard inside the container wasn't able to reach the database but changing the connection string on the appSettings.json crashes the startup and I was able to run some sql queries using sqlcmd from inside the container so I think conectivity is not the issue.

What bugs me is the difference in the logs, the first lines:

    info: OrchardCore.Environment.Shell.ShellHost[0]
      Start creation of shells
    info: OrchardCore.Environment.Shell.ShellHost[0]
      Done pre-creating and registering shells

They lack the info line of the succesfull run in localhost where the default tenant is stated to be created:

    info: OrchardCore.Environment.Shell.Builders.ShellContextFactory[0]
      Creating shell context for tenant 'Default'

**Comment by sebastienros** [comment]
The issue is that you are trying to add too many things to your startup. For instance AddOrchardCms already contains AddOrchardCore. Also you don't need to add any service for Auth as the modules will do that, even on a Blank Site recipe.

Try to follow this: https://orchardcore.readthedocs.io/en/dev/docs/guides/decoupled-cms/

**Comment by jotacv** [comment]
First of all thanks for the quick response and for the awesome software you've built ;)

I've removed the addOrchardCore call, fixed some issue with the docker entrypoint and everything seems to run fine.

The other services added (in particular the IAuthenticationProvider) are there to query ms graph on behalf of the user signed in. In particular what I needed was to access the AD groups the user belongs to in order to filter the content rendered. I've tried with the Azure Ad authentication feature and the user registration feature but without any success. So I ended up writing and injecting a custom service in the razor pages.

Another feature was to authenticate users before being able to enter the content and I dind't found any way to achieve this other than add a new AuthorizeFilter with RequireAuthenticatedUser.

**Comment by sebastienros** [comment]
Asking @MichaelPetrinolis if he would know how to solve this

**Comment by MichaelPetrinolis** [comment]
Does using OC Users with Microsoft Azure Authentication Provider and having a mechanism to map the AD Groups the user belongs to Local OC roles with the appropriate permissions solves your problem? If so, I am working on a PR and if I find some time I will complete it in the weekend.
